### ISSUE TITLE
Site Editor: adjust template parts theme paths for Atomic transfer

### ISSUE THREAD

**Issue by vindl** [Automattic/wp-calypso#50659]
Currently our FSE theme paths are prefixed with `pub/` on Dotcom (e.g. `pub/tt1-blocks`). This affects how our theme template parts taxonomies are saved, and it impacts the loading of template parts in the editor and the front end.

Everything works fine for Simple sites since correct paths are established when the site created. On the other hand, after migrating to Atomic the actual paths will be changed, since we don't have `pub/` prefix there and themes just end up in the default `wp-content/themes` folder.

To reproduce, try making some changes to template parts that will trigger CPT saving on your simple site. After transferring that site to Atomic, notice that your updates are not carried across, and that the default template part content for the theme is loaded.

**Comment by vindl** [comment]
After testing this with actual Atomic sites I confirmed that it's not causing issues with Site Editor loading template parts.

**Comment by jeyip** [comment]
I'll continually update this comment as I make progress.

## Stack Trace
- CPT templates and template parts are created through the rest templates controller ( [Link](https://github.com/WordPress/gutenberg/blob/935b05200a77dd79a80ba134ef1c794686fec4e3/lib/full-site-editing/class-wp-rest-templates-controller.php#L235-L263) )
- Before persisting data, we serialize relevant template / part information ( [Link](https://github.com/WordPress/gutenberg/blob/935b05200a77dd79a80ba134ef1c794686fec4e3/lib/full-site-editing/class-wp-rest-templates-controller.php#L328) )
- Set the CPT template / part theme taxonomy to the current theme's stylesheet ( [Link](https://github.com/WordPress/gutenberg/blob/935b05200a77dd79a80ba134ef1c794686fec4e3/lib/full-site-editing/class-wp-rest-templates-controller.php#L335-L337) )
- When creating CPTs in simple sites, the theme stylesheet will be prefixed with pub like so `pub/tt1-blocks`. This is because theme stylesheet paths are set relative to `wp-content/themes`, and dotcom nests block themes in the `wp-content/themes/pub` directory
- When migrating to atomic, themes _will not_ be nested in a pub directory. Themes will live in `wp-content/themes`, so the stylesheet will look like `tt1-blocks`
- Subsequent queries are made to the rest templates controller, which will fail for atomic sites ( [Link](https://github.com/WordPress/gutenberg/blob/935b05200a77dd79a80ba134ef1c794686fec4e3/lib/full-site-editing/class-wp-rest-templates-controller.php#L170) )
- Rest template controller helpers will query for the incorrect CPT template / part because of incorrect theme stylesheets ( [Link 1](https://github.com/WordPress/gutenberg/blob/935b05200a77dd79a80ba134ef1c794686fec4e3/lib/full-site-editing/block-templates.php#L340-L344) and [Link 2](https://github.com/WordPress/gutenberg/blob/935b05200a77dd79a80ba134ef1c794686fec4e3/lib/full-site-editing/block-templates.php#L355) )
- When the requests fail, the controller will return theme provided template files instead of CPTs ( [Link](https://github.com/WordPress/gutenberg/blob/935b05200a77dd79a80ba134ef1c794686fec4e3/lib/full-site-editing/block-templates.php#L373) )

## Possible Solution(s)
- Use an atomic hook that runs when a new site is provisioned. It can update taxonomy terms for template and template part CPTs in the database and remove existing `pub` prefixes. More on this here PCYsg-uAn-p2
  - Only run when FSE is enabled
  - Find ID for `wp_theme` taxonomy in `wp_term_taxonomy` table
  - Find all term_ids related to `wp_theme` taxonomy
  - Remove pub prefix and update each entry in the database. I may use [wp::update_db](https://developer.wordpress.org/reference/classes/wpdb/update/)
- Will also continue thinking over other possible solutions

## Outstanding Questions
- How will we handle the atomic -> simple site reversion process?
    - We can re-introduce the `pub` prefix to theme stylesheet identifiers, but I'd like to do more technical digging.
    -  It seems pretty straightforward, but I'm wondering if I'm missing any considerations 🤔 
    - Will probably split this out into a separate phab diff
    - Are template / parts created in the atomic site and their theme stylesheet associations imported back into the simple site during a lossless revert

**Comment by jeyip** [comment]
~Quick update:~

~This issue is blocked by https://github.com/Automattic/wp-calypso/issues/51246. We can't verify that our updates to the atomic transfer process fix loading issues for template / part CPTs because there is a separate problem causing fatal errors in the post and site editors. They throw a white screen of death.~

Edit: This issue is no longer blocked

**Comment by jeyip** [comment]
Created a preliminary phab diff. Not certain if it's fully functional, but I'm planning to test it tomorrow with PCYsg-3Ri-p2. Feedback for the initial draft is welcome.

**Comment by vindl** [comment]
Another thing that might need to be updated is the `post_content` of `wp_template` CPT which will contain theme data in template part attributes. Example:

```
<!-- wp:template-part {"slug":"header","theme":"tt1-blocks","tagName":"header","align":"full","className":"site-header"} /-->
```

**Comment by jeyip** [comment]
Oh that’s silly. I forgot to include a link to the Phab diff. D59225-code 😅 

> Another thing that might need to be updated is the post_content of `wp_template` CPT which will contain theme data in template part attributes. 

Thanks for the callout @vindl! I considered that and attempted to handle it in the diff. 🙂

**Comment by jeyip** [comment]
## Update
Running the async job in D59225-code and pointing it at a simple site was working. Pointing the job at an atomic site, however, was not updating database values as expected. 

This was because the methods we were using, like `wp_update_post`, were actually accessing a "cached" database on simple site infrastructure that mirrors the atomic site's database, and not the atomic site's live database itself. More context here p1616699391443600-slack-C7YPW6K40. It was suggested that we implement the logic in `wpcomsh` instead of where we handle it in D59225-code.

Also, after discussing the weird behavior for D59225-code, Grant Kinney brought up a good point that hasn't been considered yet. 
- Core WordPress has a feature that facilitates importing and exporting site content
- It probably makes sense to transform the template / part CPT references to theme stylesheets going in either direction

I made a separate issue for FSE theme pathing issues while importing / exporting content in https://github.com/Automattic/wp-calypso/issues/51442

| export | import |
| --- | --- |
| <img width="1125" alt="Screen Shot 2021-03-26 at 3 45 23 AM" src="https://user-images.githubusercontent.com/5414230/112620574-f86f4100-8de5-11eb-8868-06291eb56b3b.png">| <img width="1123" alt="Screen Shot 2021-03-26 at 3 49 44 AM" src="https://user-images.githubusercontent.com/5414230/112621054-89461c80-8de6-11eb-81a5-863d00f214c1.png"> |



## Followup Resources
- More information about the cached database and cached site infrastructure pbMoDN-Qe-p2

**Comment by jeyip** [comment]
Potential fix D60384-code that changes migration code to adjust theme paths on atomic transfer

**Comment by Copons** [comment]
@jeyip can we close this issue?

**Comment by jeyip** [comment]
> @jeyip can we close this issue?

My mistake! Don't know how I missed this for _5 months_ 😅 

https://media.giphy.com/media/xfAgILDksDJJu/giphy.gif

Yes -- we can definitely close this.
